flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.batchoperator;
+
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.ml.batchoperator.source.TableSourceBatchOp;
+import org.apache.flink.ml.common.AlgoOperator;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.util.Preconditions;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Base class of batch algorithm operators.
+ */
+public abstract class BatchOperator<T extends BatchOperator<T>> extends AlgoOperator<T> {
+
+	public BatchOperator() {
+		super();
+	}
+
+	public BatchOperator(Params params) {
+		super(params);
+	}
+
+	public <B extends BatchOperator<?>> B link(B next) {
+		return linkTo(next);
+	}
+
+	public <B extends BatchOperator<?>> B linkTo(B next) {
+		next.linkFrom(this);
```

Comment:
what's the rational of making link in this direction? what if downstream operator calls `linkFrom` twice with the same upstream; or twice with different upstreams?

Reply:
> what's the rational of making link in this direction? what if downstream operator calls `linkFrom` twice with the same upstream; or twice with different upstreams?

thx

1. with this method, the operator can link from source to sink if every operator in the dag is single input and single output and it is straightforward that execute in order from front to back
2. It is not recommended to linkFrom itself or link the same group inputs twice（added to javadoc） and the implement of the operator will define the behavior